### PR TITLE
RAW JS NFC-API

### PR DESCRIPTION
**Pull request by drandreas**
I've completed my NFC-API implementation. It's thoroughly tested and in use for 1-2 weeks. However, it contains several changes in Nordics code. I'm unsure how this can be future proofed. Moreover this rework removes `nfcURL` and `nfcRAW` in favour of full RX and TX access. `NFCon` and `NFCoff` remain unchanged.

Adding `nfcURL` and `nfcRAW` back as JS module should be straight forward. But this change still breaks backward compatibility. 

For all this reasons this pull request is intended as a basis of discussion on #1161.

### CODE REVIEW

**Comment by gfwilliams** [comment]
Thanks! I'll take a look - I wonder whether we couldn't re-add `nrfURL` to core Espruino C code, by calling your NFC functions? It might not be a huge task, and would allow Puck.js to still advertise the getting started URL on NFC out of the box.

Also, do your changes to the NFC lib work in addition of the existing functionality? I wonder whether we coild submit it back to Nordic and get them to put them in the next SDK?

**Comment by drandreas** [comment]
Yes, `nrfURL`can be re-added in custom C or also using nordics library. It just added additional code clutter and cases to handle.

While working on this project I came to the conclusion that Nordic tailored the HAL to surf the purpose of gluing the closed source NDEF library to NRF52. In SDK 13.0.0 is a second almost identical HAL, this time its tailored to NFC Tag Type 4 emulation with another closed source library.

I open some discussions within my code to illustrate both statements and give additional insights.

**Comment by drandreas** [comment]
I've re-added `NRF.nfcURL` and `NRF.nfcRaw`. I don't own an Android phone. 
Can you check if it works with yours?

**Comment by gfwilliams** [comment]
Thanks! I'll have a go - a bit busy today so can't promise anything right away though.

**Comment by gfwilliams** [comment]
This looks great - tested on my Android device and it works.

Merging now :) If you get a chance, an example in `jswrap_bluetooth.c`'s JSON showing how you can use start/stop/NFCrx/etc together to create an NFC tag would be really cool though.

**Comment by drandreas** [comment]
My forum post http://forum.espruino.com/comments/13617070/ has an example with a medium complexity. I'm not sure how to fit it properly formatted into jswrap_bluetooth.c

File: targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.c [hunk 1]
```diff
@@ -332,40 +289,20 @@ static inline void hal_nfc_common_hw_setup(uint8_t * const nfc_internal)
...
 {
-    uint8_t  nfc_internal[T2T_INTERNAL_BYTES_NR];
-    
     m_nfc_lib_callback = callback;
     m_nfc_lib_context  = p_context;
     
-    hal_nfc_common_hw_setup(nfc_internal);
+    hal_nfc_common_hw_setup(m_nfc_internal);
 
-    (void) nfc_t2t_internal_set((uint8_t *) nfc_internal, sizeof(nfc_internal));
```
[the remark this answers is not in the document]
Comment (gfwilliams) [reply targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.c]: Thanks - your method makes a lot more sense :)

File: targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.c [hunk 2]
```diff
@@ -615,45 +609,41 @@ void NFCT_IRQHandler(void)
...
-            /* Not a READ Command, so wait for next frame reception */
+            m_slp_req_received = true;
+
+            /* Wait for next frame reception */
             NRF_NFCT->TASKS_ENABLERXDATA = 1;
+
+        } else
+        if(m_nfc_lib_callback != NULL)
+        {
+            /* JS-has a high latency, set relaxed FRAMEDELAY and hope for a forgiving reader */
```
[the remark this answers is not in the document]
Comment (gfwilliams) [reply targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.c]: I wonder whether the delay could be specified along with the callback? I'm just thinking in the case of the URLs, the delay is actually quite low again now - it's only JS that's an issue as you say.

File: targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.h [hunk 3]
```diff
@@ -122,9 +158,8 @@ ret_code_t hal_nfc_start(void);
 
 /** @brief Function for sending a packet to the connected NFC reader.
   *
-  * The provided data buffer belongs to the caller and is guaranteed to be
-  * valid until the HAL_NFC_EVENT_DATA_TRANSMITTED event is received by the
-  * callback.
+  * The provided data buffer belongs to the caller and may be freed after this
```
[the remark this answers is not in the document]
Comment (gfwilliams) [reply targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.h]: It shouldn't really lead to fragmentaton, although allocating actual flat buffers can take quite a while. I guess one option might be to create a 255 byte JsVar-based buffer on `nfcStart` - so if NFC isn't being used, that data is available for other stuff.

**Comment by gfwilliams** [comment]
Thanks - I'll try and come up with something based on that

File: targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (drandreas) [reply targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.c]: I've attached figure 119 from nRF52832_PS_v1.3.pdf
figure 119
Max timeout is for a big part just a hint to the NFC peripheral’s state machine. It defines how long the peripheral should wait until it considers a missing tx command, after a rx callback, as an issue and fires an error interrupt. Therefore, as long as the rx callback initiates tx the max timeout has no effect.

However, the state machine almost always (depends on an internal state) waits after rx until max timeout is reached prior performing another rx. Hence problems may occur in cases where no response is required and sent. Quote from nRF52832_PS_v1.3.pdf: "This may happen even when the response is not required as per NFC Forum, NFC Digital Protocol Technical Specification. Any commands handled by the automatic collision resolution that don't involve a response being generated". In this such error cases the reader retries within a few ns. So keeping the timeout always at max prevents timely error recovery.

In conclusion, I would leave the implementation as it is as long as no bug reports for read errors come in. For the reasons above I also doubt it will ever hapen.

File: targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (drandreas) [reply targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.h]: I will look into that. Assuming rx will also be stored in this JVar. Could `NFCrx` be called with a view instead of a copy. In turn, speeding up request processing?

[the remark this answers is not in the document]
Comment (gfwilliams) [reply targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.h]: That sounds like a great idea! Allocating an ArrayBufferView would be super quick (it's one var), and actually would allow you to use the new queue-based stuff I added recently (you'd just push the length of received data into the queue along with BLEP_NFCRX). 

The only gotcha is the variable's contents wouldn't be guaranteed to be correct if used after event itself, but that's not like it'd be a segfault or anything.

File: targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (gfwilliams) [reply targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.c]: Sounds great - thanks!
